library/common/http/client.cc

```diff
@@ -462,13 +462,15 @@ void Client::cancelStream(envoy_stream_t stream) {
     // response code details are set on all possible paths for streams.
     direct_stream->setResponseDetails(getCancelDetails());
 
-    // The runResetCallbacks call synchronously causes Envoy to defer delete the HCM's
-    // ActiveStream. We have some concern that this could potentially race a terminal callback
-    // scheduled on the same iteration of the event loop. If we see violations in the callback
-    // assertions checking stream presence, this is a likely potential culprit. However, it's
-    // plausible that upstream guards will protect us here, given that Envoy allows streams to be
-    // reset from a wide variety of contexts without apparent issue.
-    direct_stream->runResetCallbacks(StreamResetReason::RemoteReset);
+    if (!direct_stream->callbacks_->remoteEndStreamReceived()) {
```

Comment:
Yeah this seems fine to me, though this is all too complicated to grok without digging into the code. Definitely no race condition since everything runs on one thread, just tricky sequencing of things that happen on the same event loop

Reply:
@goaway I could, but then we wouldnt' get the cancel notification up the stack.  I think if the mobile layer doesn't know the stream is done it makes sense for it to get the cancel notification, WDYT?

Reply:
As of now, the mechanism to indicate when a stream is done is by receiving a callback indicating so. This has to work systematically - this contract can't be breached.

Reply:
I think we're agreeing?  If all the data is in from the network but E-M cancels before the data is received by E-M we must call the cancel callback.  switching to ALLOW_ONLY_FOR_OPEN_STREAMS would avoid the (buggy) reset but also avoid the cancel.